src/promotion_conversion.jl

```diff
@@ -32,228 +32,55 @@ end
 
+# Change element type
+@inline Base.convert{order, dim, T1}(::Type{Tensor{order, dim, T1}}, t::Tensor{order, dim}) = Tensor{order, dim}(convert(SVector{n_components(Tensor{order, dim}), T1}, tovector(t)))
+@inline Base.convert{order, dim, T1}(::Type{SymmetricTensor{order, dim, T1}}, t::SymmetricTensor{order, dim}) = SymmetricTensor{order, dim}(convert(SVector{n_components(SymmetricTensor{order, dim}), T1}, tovector(t)))
 
-# Converting general data to a (symmetric) tensor. We leave the type of data unspecified to allow anything
-# that fulfil the contract of having a getindex and length.
-@generated function Base.convert{order, dim, T}(Tt::Union{Type{Tensor{order, dim, T}}, Type{SymmetricTensor{order, dim, T}}}, data)
-    N = n_components(get_base(get_type(Tt)))
-    return quote
-        @assert length(data) == $N
-        Tv = promote_type(T, eltype(data))
-        get_base(Tt)(to_tuple(NTuple{$N, Tv}, data))
-    end
-end
+# Peel off the M but define these so that convert(typeof(...), ...) works
+@inline Base.convert{order, dim, T1, M}(::Type{Tensor{order, dim, T1, M}}, t::Tensor{order, dim}) = convert(Tensor{order, dim, T1}, t)
+@inline Base.convert{order, dim, T1, M}(::Type{SymmetricTensor{order, dim, T1, M}}, t::SymmetricTensor{order, dim}) = convert(SymmetricTensor{order, dim, T1}, t)
+@inline Base.convert{order, dim, T1, M}(::Type{Tensor{order, dim, T1, M}}, t::SymmetricTensor{order, dim}) = convert(Tensor{order, dim, T1}, t)
+@inline Base.convert{order, dim, T1, M}(::Type{SymmetricTensor{order, dim, T1, M}}, t::Tensor{order, dim}) = convert(SymmetricTensor{order, dim, T1}, t)
 
-# Conversions to a type where the element type of the tensor is unspecified
-# calls the conversions to a type where T = eltype(data)
-Base.convert{order, dim}(Tt::Type{Tensor{order, dim}}, data) = convert(Tensor{order, dim, eltype(data)}, data)
-Base.convert{order, dim}(Tt::Type{SymmetricTensor{order, dim}}, data) = convert(SymmetricTensor{order, dim, eltype(data)}, data)
+@inline Base.convert{order, dim, T}(::Type{Tensor{order, dim}}, t::SymmetricTensor{order, dim, T}) = convert(Tensor{order, dim, T}, t)
+@inline Base.convert{order, dim, T}(::Type{SymmetricTensor{order, dim}}, t::Tensor{order, dim, T}) = convert(SymmetricTensor{order, dim, T}, t)
+@inline Base.convert{order, dim, T}(::Type{Tensor}, t::SymmetricTensor{order, dim, T}) = convert(Tensor{order, dim, T}, t)
+@inline Base.convert{order, dim, T}(::Type{SymmetricTensor}, t::Tensor{order, dim, T}) = convert(SymmetricTensor{order, dim, T}, t)
 
 # SymmetricTensor -> Tensor
 # We unroll the creation by calling the compute_index function
-@generated function Base.convert{order, dim, T1, T2, M1, M2}(::Type{Tensor{order, dim, T1, M1}}, t::SymmetricTensor{order, dim, T2, M2})
+@generated function Base.convert{order, dim, T1, T2}(::Type{Tensor{order, dim, T1}}, t::SymmetricTensor{order, dim, T2})
     N = n_components(Tensor{order, dim})
     rows = Int(N^(1/order))
```

Comment:
`N` on the line above is only used to calculate `rows`
remove those lines and change `rows` to `dim` in the loops!